### PR TITLE
[ObjCARC] Add variations of intrinsics that do not touch reference counts

### PR DESCRIPTION
**Pull request by AZero13**
memset, memmove, and memcpy have variations that are not covered by the switch case but should be. They do not touch the retain count either.

### CODE REVIEW

**Comment by fhahn** [comment]
Could you add a test?

**Comment by AtariDreams** [comment]
> Could you add a test?

I do not know how to test this. I just know that when this function was first written, these intrinsics were not part of LLVM. And there is no reason to believe memcpy will not change retain counts but memcpy_inline will.

**Comment by AtariDreams** [comment]
@fhahn Please help me. How would I test this?

**Comment by AtariDreams** [comment]
@rjmccall Thoughts on this?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -4241,9 +4241,20 @@ SelectionDAG::computeOverflowForSignedAdd(SDValue N0, SDValue N1) const {
   // cannot overflow.
   if (ComputeNumSignBits(N0) > 1 && ComputeNumSignBits(N1) > 1)
     return OFK_Never;
+  
+    // smulhi + any value never overflow
+  KnownBits N1Known = computeKnownBits(N1);
+  if (N0.getOpcode() == ISD::SMUL_LOHI)
+    return OFK_Never;
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4248]: Unrelated change

**Review by arsenm — COMMENTED** [review]
Missing tests. Also the whole function could probably be expressed in terms of attributes

**Comment by rjmccall** [comment]
The test would probably be something like removing a retain/release pair on opposite sides of an intrinsic call.

**Comment by AtariDreams** [comment]
> The test would probably be something like removing a retain/release pair on opposite sides of an intrinsic call.

Thank you. Okay, I think I know what to do now.
